grpcx/deamon.go

```diff
@@ -0,0 +1,84 @@
+package grpcx
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"google.golang.org/grpc"
+
+	"github.com/socialpoint-labs/bsk/contextx"
+)
+
+// Daemon represents a runnable gRPC daemon
+type Daemon struct {
+	port int
+	lis  net.Listener
+	svr  *grpc.Server
+	ef   func(error)
+	apps []Application
+}
+
+// Application is a contextx.Runner that can register gRPC services
+type Application interface {
+	contextx.Runner
+	RegisterGRPC(grpc.ServiceRegistrar)
+}
+
+// NewDaemonWithPort creates a gRPC Daemon listening at the given port.
+func NewDaemonWithPort(port int, svr *grpc.Server, ef func(error), apps ...Application) Daemon {
+	return Daemon{
+		port: port,
+		svr:  svr,
+		ef:   ef,
+		apps: apps,
+	}
+}
+
+// NewDaemon creates a gRPC Daemon for the given listener.
+//
+// The lifecycle of the listener will *not* be managed by this function, nor by the Daemon. The application
+// needs to take manage it.
+func NewDaemon(lis net.Listener, svr *grpc.Server, ef func(error), apps ...Application) Daemon {
+	return Daemon{
+		lis:  lis,
+		svr:  svr,
+		ef:   ef,
+		apps: apps,
+	}
+}
+
+// Run registers every application of the Daemon with a gRPC server and runs them,
+// then starts the server and gracefully shuts it down once the context is done.
+func (d Daemon) Run(ctx context.Context) {
+	for _, app := range d.apps {
+		app.RegisterGRPC(d.svr)
+
+		go app.Run(ctx)
+	}
+
+	listener, err := d.listener()
+	if err != nil {
+		d.ef(err)
```

Comment:
why the function and no the returning the error directly?

Reply:
Here we're adhering to the contextx.Runner interface, which is the type of our runners (see https://github.com/socialpoint-labs/bsk/blob/master/contextx/runner.go#L8). 

As the name implies, the runner is considered "something you can run" typically from bootstrapping the application, but that is self-managed. This allows us to treat them all the same and to manage their lifecycle in a generic way. Here you can see how they're started: https://github.com/socialpoint/backend-config-service/pull/95/files#diff-2cb0e862e8191622e33c69b6920ad67c6ba5614502a7469937330f7328d7f184R59

Reply:
> Here we're adhering to the contextx.Runner interface, which is the type of our runners
This is a good point...

> As the name implies, the runner is considered "something you can run" typically from bootstrapping the application, but that is self-managed.
But even in that way you can allow to return an error and the client on his own application can handle that error in the way that they want, with your implementation you are adding a new step where they need a function to handle that errors

But if you are commenting, we are using an interface that directly can't return an error, and do a refactor of that could be a mess we can maintain as your proposed